woocommerce/woocommerce-gateway-stripe
# Add a filter call when updating an existent intent (saved in the order)

Fixes #2240

## Changes proposed in this Pull Request:

Currently there is a filter call when creating a new intent (`wc_stripe_generate_create_intent_request`), but there is no filter when updating an existent intent.

This PR adds a new filter call when updating an existent intent (`wc_stripe_update_existing_intent_request`)  in `update_existing_intent()`

**Merchant's use case:**
> We sell weighed goods where the exact weight is not known until order fulfillment. Our integration displays messages to the customer on the product pages and checkout page explaining that a buffer of X% will be held on their card to account for exact weighing. We add the exact weights during fulfillment, which updates the price and then we can finish the charge (having already authorized a sufficient buffer).

The merchant is currently using the filter `wc_stripe_generate_create_intent_request` in the function `generate_create_intent_request()` to add the buffer when the intent is created.
But when the payment fails and the user retries, the intent is read from the order and `update_existing_intent()` is called instead of `generate_create_intent_request()`... and in this update function the amount is recalculated and set to the request:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fe881c4c271fcbe04574aa01c8362558d982150c/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1357-L1360

So when the payment is retried the order amount is used, not the one modified in the `wc_stripe_generate_create_intent_request` filter, and there is no filter available in `update_existing_intent()` to change it. 

Additional context:
- 4609846-zd-woothemes
- p1639438176255700-slack-C7U3Y3VMY

## Testing instructions

1. Install the [Code Snippets](https://wordpress.org/plugins/code-snippets) plugin.
2. Go to **Snippets** and add a **New Snippet** with this content:
```php
function example_callback( $request, $order, $prepared_source ) {
    $request['amount'] = $request['amount'] + 100;
    return $request;
}
add_filter( 'wc_stripe_update_existing_intent_request', 'example_callback', 10, 3 );
```
3. Save changes and activate it.
4. Go to the Store and add any product to the cart.
5. Go to the checkout page.
6. Use the card 4000002500003155 and intentionally FAIL the 3DS validation.
7. Click `Place order` again, but this time complete the authentication.
8. Go to the [Stripe dashboard](https://dashboard.stripe.com/test/payments), and verify that the amount of the payment for the purchased order, includes the $100 buffer added in the filter in `Step 2`.
